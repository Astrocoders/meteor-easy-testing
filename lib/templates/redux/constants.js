export default ({ name }: { name:string }):string => (
`export default ACTION = '${name.toLowerCase()}/ACTION'`
)
