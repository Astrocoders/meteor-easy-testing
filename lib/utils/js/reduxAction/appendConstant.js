import j from 'jscodeshift'

export default function appendConstant({ast, name, value}) {
  const newVariableDeclaration = j.variableDeclaration('const', [
    j.variableDeclarator(j.identifier(name), j.literal(value)),
  ])
  const newExport = j.exportNamedDeclaration(newVariableDeclaration, [], null)
  const exports = ast
    .find(j.ExportNamedDeclaration)

  return exports
    .at(exports.length - 1)
    .insertAfter(newExport)
}
