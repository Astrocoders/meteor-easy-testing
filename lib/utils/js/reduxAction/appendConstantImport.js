import j from 'jscodeshift'

const CONSTANTS_IMPORTED_FROM = './constants'

export default function appendConstantImport({ast, constantName}) {
  const newImportSpecifier = j.importSpecifier(j.identifier(constantName), j.identifier(constantName))
  const importeds = ast
    .find(j.ImportDeclaration, node => node.source.value === CONSTANTS_IMPORTED_FROM)
    .find(j.ImportSpecifier)

  importeds
    .at(importeds.nodes().length - 1)
    .insertAfter(newImportSpecifier)
}
