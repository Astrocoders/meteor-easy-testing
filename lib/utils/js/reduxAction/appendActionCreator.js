import j from 'jscodeshift'
import appendConstantImport from './appendConstantImport'

export default function appendActionCreator({ast, constantName, actionCreatorName}) {
  appendConstantImport({
    ast,
    constantName,
  })

  appendActionCreatorFunction({
    ast,
    constantName,
    actionCreatorName,
  })

  return ast
}

function appendActionCreatorFunction({ast, constantName, actionCreatorName}) {
  const newFunction = j.functionDeclaration(j.identifier(actionCreatorName), [], j.blockStatement([
    j.returnStatement(
      j.objectExpression([
        j.property('init', j.literal('type'), j.identifier(constantName)),
      ])
    ),
  ]))
  const newExport = j.exportNamedDeclaration(newFunction, [], null)
  const exports = ast
    .find(j.ExportNamedDeclaration)

  exports
    .at(exports.length - 1)
    .insertAfter(newExport)
}
