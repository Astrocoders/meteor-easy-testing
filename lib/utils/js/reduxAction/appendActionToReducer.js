import j from 'jscodeshift'
import appendConstantImport from './appendConstantImport'

export default function appendActionToReducer({ ast, constantName }){
  appendConstantImport({ ast, constantName })

  const switchCase = j.switchCase(j.identifier(constantName), [
    j.returnStatement(j.identifier('state')),
  ])

  return ast
  	.find(j.SwitchCase, node => node.test === null)
  	.insertBefore(switchCase)
}
