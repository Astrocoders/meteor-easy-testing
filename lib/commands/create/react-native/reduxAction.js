import path from 'path'
import j from 'jscodeshift'
import { snakeCase, camelCase } from 'lodash/fp'
import pathes from '~/utils/pathes'
import appendConstant from '~/utils/js/reduxAction/appendConstant'
import appendActionCreator from '~/utils/js/reduxAction/appendActionCreator'
import appendActionToReducer from '~/utils/js/reduxAction/appendActionToReducer'
import readFile from '~/utils/filesystem/readFile'
import saveASTtoFile from '~/utils/js/saveASTtoFile'

export default function reduxAction({ name: definition }){
  const entity = path.dirname(definition)
  const name = path.basename(definition)
  const entityPath = path.join(pathes.redux, entity)
  const actionsFilePath = path.join(entityPath, 'actions.js')
  const reducerFilePath = path.join(entityPath, 'reducer.js')
  const constantsFilePath = path.join(entityPath, 'constants.js')
  const actionsFile = readFile({filePath: actionsFilePath})
  const reducerFile = readFile({filePath: reducerFilePath})
  const constantsFile = readFile({filePath: constantsFilePath})

  const constantName = snakeCase(name).toUpperCase()

  // Include new constant
  saveASTtoFile({
    filePath: constantsFilePath,
    ast: appendConstant({
      ast: j(constantsFile),
      name: constantName,
      value: `${camelCase(entity)}/${constantName}`,
    }),
  })

  // Add new action creator
  saveASTtoFile({
    filePath: actionsFilePath,
    ast: appendActionCreator({
      ast: j(actionsFile),
      constantName,
      actionCreatorName: name,
    }),
  })

  // Add new action creator
  saveASTtoFile({
    filePath: reducerFilePath,
    ast: appendActionToReducer({
      ast: j(reducerFile),
      constantName,
    }),
  })
}
