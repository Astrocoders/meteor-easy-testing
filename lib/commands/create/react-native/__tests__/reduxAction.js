import { stripIndents } from 'common-tags'
import reduxAction from '../reduxAction'

const mockedActionsFile = stripIndents`
import { FOO } from './constants'
export function foo(){
  return {
    type: FOO,
  }
}
`

const mockedReducerFile = stripIndents`
import { FOO } from './constants'

export default function mocked(state, {type, payload}){
  switch(type){
    case FOO: return {...state}
    default: return {...state}
  }
}
`

const mockedConstantsFile = `export const FOO = 'mocked/FOO'`

jest.mock('~/utils/js/saveASTtoFile', () => jest.fn())
jest.mock('~/utils/filesystem/readFile', () => (
  jest.fn()
    .mockImplementation(({ filePath }) => {
      if(filePath.includes('action')) return mockedActionsFile
      if(filePath.includes('reducer')) return mockedReducerFile
      if(filePath.includes('constants')) return mockedConstantsFile
    })
))

describe('reduxActions', () => {
  const definition = 'Mocked/BAR'
  const saveASTtoFile = require('~/utils/js/saveASTtoFile')

  beforeAll(() => {
    reduxAction({ name: definition })
  })

  test('it should properly change the constants file', () => {
    const constantsChangeCall = saveASTtoFile.mock.calls[0][0]

    expect(constantsChangeCall).toBeDefined()
    expect({
      filePath: constantsChangeCall.filePath,
      ast: constantsChangeCall.ast.toSource({})
    }).toMatchSnapshot()
  })

  test('it should properly change the actions file', () => {
    const actionsChangeCall = saveASTtoFile.mock.calls[1][0]

    expect(actionsChangeCall).toBeDefined()
    expect({
      filePath: actionsChangeCall.filePath,
      ast: actionsChangeCall.ast.toSource({})
    }).toMatchSnapshot()
  })

  test('it should properly change the reducers file', () => {
    const reducersChangeCall = saveASTtoFile.mock.calls[2][0]

    expect(reducersChangeCall).toBeDefined()
    expect({
      filePath: reducersChangeCall.filePath,
      ast: reducersChangeCall.ast.toSource({})
    }).toMatchSnapshot()
  })
})

