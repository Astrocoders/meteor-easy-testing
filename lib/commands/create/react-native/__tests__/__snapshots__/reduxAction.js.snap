// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduxActions it should properly change the actions file 1`] = `
Object {
  "ast": "import {FOO, BAR} from './constants';
export function foo(){
return {
type: FOO,
}
}

export function BAR() {
  return {
    \\"type\\": BAR
  };
}",
  "filePath": "src/redux/Mocked/actions.js",
}
`;

exports[`reduxActions it should properly change the constants file 1`] = `
Object {
  "ast": "export const FOO = 'mocked/FOO'
export const BAR = \\"mocked/BAR\\";",
  "filePath": "src/redux/Mocked/constants.js",
}
`;

exports[`reduxActions it should properly change the reducers file 1`] = `
Object {
  "ast": "import {FOO, BAR} from './constants';

export default function mocked(state, {type, payload}){
switch (type) {
case FOO: return {...state}
case BAR:
  return state;
default: return {...state}
}
}",
  "filePath": "src/redux/Mocked/reducer.js",
}
`;
